### DISCUSSION TITLE
RegisterInstance for reuse in current scope

### DISCUSSION THREAD

**Discussion by timrhaynes** [dadhi/DryIoc#411] [Q&A]
In DryIoc 2.12.10 (old, I know) the RegisterInstance method had a reuse parameter where you could indicate Reuse.InCurrentScope.

In DryIoc 4.7.5, that parameter is gone. Is there a new way to Register an instance for reuse in the current scope?

Thanks

**Comment by dadhi** [comment]
@timrhaynes Use the `Use` method, no joking.

**Reply by timrhaynes** [reply]
Thank you for the suggestion, but I couldn't get that to work. It kept throwing errors when trying to instantiate a controller. For your interest, here is that error. It seems to be only looking in the new scope but ignoring the parent scope, which I know doesn't make any sense given the error message:

> code: Error.UnableToResolveUnknownService;
> message: Unable to resolve resolution root Microsoft.AspNetCore.Mvc.Controller {ServiceKey="HomeController"}
>   from container without scope
>  with Rules with {TrackingDisposableTransients, SelectLastRegisteredFactory} and without {ThrowIfDependencyHasShorterReuseLifespan, ThrowOnRegisteringDisposableTransient}
>  with FactorySelector=SelectLastRegisteredFactory
>  with Made={FactoryMethod=ConstructorWithResolvableArguments}
> Where no service registrations found
>   and no dynamic registrations found in 0 of Rules.DynamicServiceProviders
>   and nothing found in 0 of Rules.UnknownServiceResolvers

How would you switch this to Use? Thank you!

```
 public void RegisterInstanceInScope<T>(T instance, string serviceName = null)
        {
            if (!string.IsNullOrEmpty(serviceName))
            {
                _internalDryContainer.RegisterInstance(instance: instance, reuse: Reuse.InCurrentScope, serviceKey: serviceName, ifAlreadyRegistered: IfAlreadyRegistered.Replace);
            }
            else
            {
                _internalDryContainer.RegisterInstance(instance: instance, reuse: Reuse.InCurrentScope, ifAlreadyRegistered: IfAlreadyRegistered.Replace);
            }
        }
```

**Reply by dadhi** [reply]
`Use` does not support service names, so like that:

```cs
scopedContainer.Use<X>(x)
```

**Reply by timrhaynes** [reply]
Since I may need to use service names to register an instance, and therefore cannot use `Use`, I guess I have to go back to my original question of:

> Is there a new way to Register an instance for reuse in the current scope?

Thanks again!

**Reply by dadhi** [reply]
Then try `RegisterDelegate<X>(_ => x, Reuse.Scoped, IfAlreadyRegistered.Replace)`

**Reply by timrhaynes** [reply]
I will try that, thank you!
